library/net_sockets.c

```diff
@@ -216,6 +216,128 @@ int mbedtls_net_connect( mbedtls_net_context *ctx, const char *host,
     return( ret );
 }
 
+/*
+ * Initiate a TCP connection with host:port and the given protocol with a timeout (ms)
+ */
+int mbedtls_net_connect_timeout( mbedtls_net_context *ctx, const char *host,
+                         const char *port, int proto, int timeout )
+{
```

Comment:
Most of the code is copied from `mbedtls_net_connect`. Code duplication is bad for maintenance (and also wastes code size). Please change `mbedtls_net_connect` to be a backward compatibility wrapper for calling the new function with timeout=0.